Make linenoiseState a class with methods, various other adjustments

starseeker opened the pull request:
Follow on pull request from https://github.com/yhirose/cpp-linenoise/pull/33 that attempts to minimize the formatting changes.  Also attempts to add a way to keep the old API (or at least the key parts of it) working - both the old version of the example and the updated version have been compiled against it.

Note that, unlike that PR, this one does not incorporate the https://github.com/yhirose/cpp-linenoise/pull/32 fixes - those are still of interest, but since that PR is still open we'll skip pulling it in here.

Might still be a few more tweaks to make (method name cleanups, etc.) but submitting in its current form to get things moving.

starseeker committed:
Reduce variable scope

starseeker committed:
Make most linenoise functions into methods on a class.

starseeker committed:
Support previous API style

Implement a way to allow the previous API as seen in the old example.cpp to
keep working.

starseeker committed:
Add a mutex to protect RefreshLine if multiple threads need to call it.

starseeker committed:
Adjust how history and its up/down arrow browsing are handled.

starseeker committed:
We don't want to hang if we can't read anything from stdin

starseeker committed:
Add WipeLine and ClearScreen methods.

starseeker committed:
See if we can avoid using strcasecmp

starseeker committed:
Correct behavior for sending commands

Fix from Chris McGregor (f4alt <christopherjmacgregor@gmail.com>:
Old linenoise behavior would send commands on newline. For subprocess use, like
we do with nirt in MGED, this is very important as the alternative means
collecting commands like "Command_0\nCommnand_1.." and expecting the process to
parse the newlines.

starseeker committed:
Misc. fixes and adjustments

starseeker commented:
@yhirose A (somewhat belated) thought - if you're not really looking to change cpp-linenoise further, would it be better to set up a fork project and use that to pull in changes and do new work?  Don't want to shove stuff at you that wouldn't be a benefit...

starseeker committed:
Compare against the correct string

yhirose commented:
> @yhirose A (somewhat belated) thought - if you're not really looking to change cpp-linenoise further, would it be better to set up a fork project and use that to pull in changes and do new work? Don't want to shove stuff at you that wouldn't be a benefit...

Don't worry about it. Any improvements are welcome. :)

yhirose commented on linenoise.hpp:
```diff
@@ -1573,20 +1641,29 @@ inline int unicodeReadUTF8Char(int fd, char* buf, int* cp)
 /* Return true if the terminal name is in the list of terminals we know are
  * not able to understand basic escape sequences. */
-inline bool isUnsupportedTerm(void) {
+static bool isUnsupportedTerm(void) {
```
Is there any reason to change it to `static` from `inline`?

starseeker replied:
Not really - I guess I'm just more used to static.  I'll switch it back.

yhirose commented on linenoise.hpp:
```diff
@@ -1763,12 +1855,12 @@ inline void linenoiseBeep(void) {
  * The state of the editing is encapsulated into the pointed linenoiseState
  * structure as described in the structure definition. */
-inline int completeLine(struct linenoiseState *ls, char *cbuf, int *c) {
+int linenoiseState::completeLine(char *cbuf, int *c) {
```
If you remove `inline` keyword, we are not able to include the `linenoise.hpp` in more than one compilation unit. For example, if we include this file in `a.cpp` and `b.cpp` and link them to make an executable, the linker doesn't allow use to build.

starseeker replied:
Got it - I pushed a commit adding them back in, as well as an example to build that tests that inclusion case.

yhirose commented on linenoise.hpp:
```diff
@@ -145,22 +145,112 @@
+        std::string history_tmpbuf;
+
+        size_t history_max_len = LINENOISE_DEFAULT_HISTORY_MAX_LEN;
+        std::vector<std::string> history;
```
Is it possible to suffix all private member variables with an underscore, such as len_, for better code readability in method bodies? Otherwise, it is quite difficult for me to distinguish whether they are member variables or local variables.

starseeker replied:
I added an l_ prefix - will that do the trick?

yhirose replied:
I actually prefer `_` suffix and using the style in my other projects. :)
https://micro-os-plus.github.io/develop/naming-conventions/#private-member-variables-names

starseeker replied:
OK, one second...

starseeker replied:
Switched l_ names to using the _ suffix instead.

yhirose commented on linenoise.hpp:
```diff
@@ -1940,151 +2031,166 @@ inline void refreshMultiLine(struct linenoiseState *l) {
-        strcpy(l->buf,history[history.size() - 1 - l->history_index].c_str());
-        l->len = l->pos = static_cast<int>(strlen(l->buf));
-        refreshLine(l);
+void linenoiseState::linenoiseEditHistoryNext(int dir) {
```
The new history handling code looks very different from the original linenoise code of antires. Does it have the same behavior? If not, how does it work?

starseeker replied:
It's similar but probably not quite identical.  I was trying to make it easier to reason about - the "working" buffer is stashed in a temporary buffer rather than the history vector, since its contents (unlike the rest of the history) may change up until the line is entered.  Also, we avoid the need for the initial "AddHistory("")" which felt a little strange to me - in the current approach the indexing and data reflect the fact that there *is* no history until something is entered.

yhirose replied:
If the new behavior is different from the original, could you please put it in another pull request? This way, we can ensure that this pull request doesn't cause any behavior changes. I would like to handle things one at a time. Thanks for your understanding!

starseeker replied:
OK, I've backed them out.

yhirose commented:
@starseeker I just quickly walked through the changes, and they basically look very good to me!

I left some comments regarding the followings:
1. Missing `inline` keyword before some free function bodies and instance method bodies.
2. Naming of private member variables of `linenoiseState`
3. The history behavior

starseeker committed:
Use inline

starseeker committed:
Restore inlines to allow multiple includes

Also add a compilation test case to examples to verify we can include more than
once and create an executable.

starseeker committed:
Add an l_ prefix to all class member variables.

starseeker committed:
Stray static global mess with history rename.  Add some comments.

starseeker committed:
Eliminate a couple more globals

starseeker committed:
Back out history changes

starseeker committed:
Use suffix character for class variables

starseeker committed:
Remove unnecessary prefixes from internal class method names

starseeker commented:
@yhirose I think that's got it?  Let me know if there's anything else.  I'll look again at the history changes and submit a separate pull request if I can confirm whether there was a functional implication there.

starseeker committed:
Clean up class API slightly.

starseeker committed:
Unused define

yhirose commented:
@starseeker looks very good! Just one more change please. Could you rename `main.cpp` `f1.cpp` `f2.cpp` to `multiinclude_main.cpp` `multiinclude_f1.cpp` `multiinclude_f2.cpp`?

EDIT: `multiinclude/main.cpp` `multiinclude/f1.cpp` `multiinclude/f2.cpp` are actually better.

starseeker committed:
Move multiinclude test files to subdir

starseeker commented:
All set - moved to subdirectory.

yhirose commented:
@starseeker thanks for your hard work!

gyulamad commented:
wow. that was quick! thank you guys :) 
@yhirose @starseeker
